[libc] Implement more input functions on the GPU

jhuber6 opened the pull request:
Summary:
This patch implements the `fgets`, `getc`, `fgetc`, and `getchar`
functions on the GPU. Their implementations are straightforward enough.
One thing worth noting is that the implementation of `fgets` will be
extremely slow due to the high latency to read a single char. A faster
solution would be to make a new RPC call to call `fgets` (due to the
special rule that newline or null breaks the stream). But this is left
out because performance isn't the primary concern here.

jhuber6 committed:
[libc] Implement more input functions on the GPU

Summary:
This patch implements the `fgets`, `getc`, `fgetc`, and `getchar`
functions on the GPU. Their implementations are straightforward enough.
One thing worth noting is that the implementation of `fgets` will be
extremely slow due to the high latency to read a single char. A faster
solution would be to make a new RPC call to call `fgets` (due to the
special rule that newline or null breaks the stream). But this is left
out because performance isn't the primary concern here.

jhuber6 committed:
Add support docs

sivachandra reviewed (approved):
OK for non-GPU changes.

jhuber6 committed:
Fix clearerr leftover code

jhuber6 committed:
Make new ferror functions work when called with standard streams

Summary:
When crossing the device boundary we will find that the standard stream
have different pointers. This is intentional as it allows us to perform
IO on the standard streams without first needing to do any setup.
However, this needs to be accounted for in the implementation. This
patch adds some logic to  encode whether or not this is a standard
stream in the low orders of the bits. Because these are aligned to at
least a 4 bytes boundary we will always have access to 2 bits to store 4
different values.
